bot/main.py

```diff
@@ -0,0 +1,811 @@
+        if not signature:
+            await _error_response(send, b"Missing signature header")
+            return
+
+        more_body = True
+        payload = bytearray()
+        while more_body:
+            event = await receive()
+            match event:
+                case {"type": "http.request"}:
+                    more_body = event.get("more_body", False)
+                    payload.extend(event.get("body") or b"")
+                case _:
+                    raise NotImplementedError
+
+        digest = "sha256=" + hmac.new(CLIENT_SECRET, payload, digestmod="sha256").hexdigest()
+
+        if not hmac.compare_digest(signature.decode(), digest):
+            await _error_response(send, b"Invalid signature", status_code=401)
+            return
+
+        await self.app(scope, _CachedReceive(payload, receive), send)
+
+
+@dataclasses.dataclass(slots=True)
+class _Workflow:
+    name: str
+    workflow_id: int
+
+
+class _WorkflowAction(str, enum.Enum):
+    COMPLETED = "completed"
+    IN_PROGRESS = "in_progress"
+    REQUESTED = "requested"
+
+
+class _WorkflowDispatch:
+    """Workflow dispatch tracker."""
+
+    __slots__ = ("_listeners",)
+
+    def __init__(self) -> None:
+        self._listeners: dict[
+            tuple[int, int, str], streams.MemoryObjectSendStream[tuple[int, str, _WorkflowAction]]
+        ] = {}
+
+    def consume_event(self, body: dict[str, typing.Any], /) -> None:
+        """Dispatch a workflow_run event.
+
+        Parameters
+        ----------
+        body
+            Dict body of the workflow_run event.
+        """
+        head_repo_id = int(body["workflow_run"]["head_repository"]["id"])
+        head_sha = body["workflow_run"]["head_sha"]
+        repo_id = int(body["repository"]["id"])
+
+        if send := self._listeners.get((repo_id, head_repo_id, head_sha)):
+            action = _WorkflowAction(body["action"])
+            name = body["workflow_run"]["name"]
+            workflow_id = int(body["workflow_run"]["id"])
+            send.send_nowait((workflow_id, name, action))
+
+    @contextlib.contextmanager
+    def track_workflows(
+        self, repo_id: int, head_repo_id: int, head_sha: str, /
+    ) -> collections.Generator["_IterWorkflows", None, None]:
+        """Async context manager which manages tracking the workflows for a PR.
+
+        Parameters
+        ----------
+
+        Returns
+        -------
+        _IterWorkflows
+            Async iterable of the received workflow finishes.
+
+            `_IterWorkflows.filter_names` should be used to set this to filter
+            for specific names before iterating over this.
+        """
+        # TODO: the typing for this function is wrong, we should be able to just pass item_type.
+        key = (repo_id, head_repo_id, head_sha)
+        send, recv = anyio.create_memory_object_stream(math.inf, item_type=tuple[int, str, _WorkflowAction])
```

Comment:
Set a sanity check ceiling like 1k where if it errors then something is obv really broken instead of inf